flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalConnectionHandlingTest.java

```diff
@@ -291,55 +271,72 @@ private void writeLeaderInformationToZooKeeper(
+        testWithQueueLeaderElectionListener(
+                queueLeaderElectionListener ->
+                        new ZooKeeperLeaderRetrievalDriver(
+                                zooKeeperClient,
+                                "/testNewLeaderAfterReconnectTriggersListenerNotification",
+                                queueLeaderElectionListener,
+                                ZooKeeperLeaderRetrievalDriver.LeaderInformationClearancePolicy
+                                        .ON_SUSPENDED_CONNECTION,
+                                fatalErrorHandlerResource.getFatalErrorHandler()),
+                (leaderRetrievalDriver, queueLeaderElectionListener) -> {
+                    final String leaderAddress = "foobar";
+                    final UUID sessionId = UUID.randomUUID();
+                    writeLeaderInformationToZooKeeper(
+                            leaderRetrievalDriver.getConnectionInformationPath(),
+                            leaderAddress,
+                            sessionId);
+
+                    // pop new leader
+                    queueLeaderElectionListener.next();
+
+                    testingServer.stop();
+
+                    final CompletableFuture<String> connectionSuspension =
+                            queueLeaderElectionListener.next();
+
+                    // wait until the ZK connection is suspended
+                    connectionSuspension.join();
+
+                    testingServer.restart();
+
+                    final String newLeaderAddress = "barfoo";
+                    final UUID newSessionId = UUID.randomUUID();
+                    writeLeaderInformationToZooKeeper(
+                            leaderRetrievalDriver.getConnectionInformationPath(),
+                            newLeaderAddress,
+                            newSessionId);
+
+                    // check that we find the new leader information eventually
+                    CommonTestUtils.waitUntilCondition(
+                            () -> {
+                                final CompletableFuture<String> afterConnectionReconnect =
+                                        queueLeaderElectionListener.next();
+                                return afterConnectionReconnect.get().equals(newLeaderAddress);
+                            },
+                            Deadline.fromNow(Duration.ofSeconds(30L)));
+                });
+    }
+
+    private void testWithQueueLeaderElectionListener(
+            FunctionWithException<
+                            QueueLeaderElectionListener, ZooKeeperLeaderRetrievalDriver, Exception>
+                    driverFactoryMethod,
+            BiConsumerWithException<
+                            ZooKeeperLeaderRetrievalDriver, QueueLeaderElectionListener, Exception>
+                    testCallback)
+            throws Exception {
         final QueueLeaderElectionListener queueLeaderElectionListener =
```

Comment:
Why not model this as an extension that can be reused more easily?

Reply:
Good point 👍